src/main.c

```diff
@@ -0,0 +1,216 @@
+        .id_mask = CAN_STD_ID_MASK
+};
+
+//filter-id=0x456
+const struct zcan_filter single_filter_2 =
+ {
+        .id_type = CAN_STANDARD_IDENTIFIER,
+        .rtr = CAN_DATAFRAME,
+        .id = 0x456,
+        .rtr_mask = 1,
+        .id_mask = CAN_STD_ID_MASK
+};
+
+//filter-mask =0 (essentially ignores the id)
+const struct zcan_filter all_filter =
+ {
+        .id_type = CAN_STANDARD_IDENTIFIER,
+        .rtr = CAN_DATAFRAME,
+        .id = 0x123,
+        .rtr_mask = 0,
+        .id_mask = CAN_STD_ID_MASK
+};
+
+
+//CAN messages
+const struct device *can_dev;
+int ret;
+int ret1;
+int filter_id1;
+int filter_id2;
+
+//LED
+const struct device *led_dev;
+const struct device *led_dev2;
+const struct device *led_dev3;
+const struct device *led_dev4;
+bool led_status;
+
+
+//send a single message to bus
+void send_message(struct zcan_frame *frame)
+{
+        ret = can_send(can_dev, frame, K_MSEC(150), NULL, NULL); // send message 1
+        if (ret != CAN_TX_OK) {
+                printk("Sending failed [%d]", ret);
+                }         
+}
+
+//send 2 messages to bus
+void send_message_2(void)
+{
+        ret = can_send(can_dev, &frame1, K_MSEC(100), NULL, NULL); // send message 1
+        if (ret != CAN_TX_OK) {
+                printk("Sending failed [%d]", ret);
+                }     
+        k_sleep(K_MSEC(250)); //wait before sending again
+
+        ret1 = can_send(can_dev, &frame2, K_MSEC(100), NULL, NULL); //send message 2
+        if (ret1 != CAN_TX_OK) {
+                printk("Sending failed [%d]", ret);
+                }     
+
+}
+
+//recieves all messages using filter with ID mask = 0
+void rx_callback_function_no_filter(struct zcan_frame *frame, void *arg)
+{       
+        //turn on Red LED when messages are recived using this callback function
+        gpio_pin_set(led_dev, PIN0, (int)true);
+}
+
+//recieves 1 message using filter for message id and blink LED if received
+void rx_callback_function_single_filter(struct zcan_frame *frame, void *arg)
+{
+        //turn on Orange LED when message received matches filter criteria
+        gpio_pin_set(led_dev2, PIN1, (int)true);
+}
+
+
+//receives messages using a message queue for all messages
+void rx_callback_functionq(struct k_msgq *message_q) //Activity 3
+{
+        //turn on  LED when message received matches filter criteria
+        gpio_pin_set(led_dev3, PIN2, (int)true);
+}
+
+
+//Receiver for worker queue
+void receiver_q(struct zcan_filter filter)
+{
+        struct zcan_frame message;
+        int filter_id = can_attach_msgq(can_dev, &receive_q, &filter);
+        while(1)
+        {
+                k_msgq_get(&receive_q, &message, K_FOREVER);
+
+                if (message.dlc != 2U) {
+			printk("Wrong data length: %u\n", message.dlc);
+			continue;
+		}
+                //turn on  LED when message received matches filter criteria
+                gpio_pin_set(led_dev4, PIN3, (int)true);
+        }
+}
+
+//Setup LED for recieve
+void led_setup()
```

Comment:
This is great to help separate the work into clear groups and make code easier to maintain.